monkey/monkey_island/cc/repository/mongo_node_repository.py

```diff
@@ -0,0 +1,87 @@
+from copy import deepcopy
+from typing import Any, MutableMapping, Sequence
+
+from pymongo import MongoClient
+
+from monkey_island.cc.models import CommunicationType, MachineID, Node
+
+from . import INodeRepository, RemovalError, RetrievalError, StorageError
+from .consts import MONGO_OBJECT_ID_KEY
+
+UPSERT_ERROR_MESSAGE = "An error occurred while attempting to upsert a node"
+SRC_FIELD_NAME = "machine_id"
+
+
+class MongoNodeRepository(INodeRepository):
+    def __init__(self, mongo_client: MongoClient):
+        self._nodes_collection = mongo_client.monkey_island.nodes
+
+    def upsert_communication(
+        self, src: MachineID, dst: MachineID, communication_type: CommunicationType
+    ):
+        try:
+            node_dict = self._nodes_collection.find_one({SRC_FIELD_NAME: src})
+        except Exception as err:
+            raise StorageError(f"{UPSERT_ERROR_MESSAGE}: {err}")
+
+        if node_dict is None:
+            updated_node = Node(machine_id=src, connections={dst: frozenset((communication_type,))})
+        else:
+            node = MongoNodeRepository._mongo_record_to_node(node_dict)
+            updated_node = MongoNodeRepository._add_connection_to_node(
+                node, dst, communication_type
+            )
```

Comment:
If we defined a Node document, saving a Node pydantic object **should** be as simple as `NodeDocument(**updated_node.dict()).save()`

Reply:
Reverse should also be as simple as `Node.parse(node_document.to_mongo())`

Reply:
Can we create the NodeDocument automatically from the pydantic object without adding more boilerplate? Will we still need to remove the mongo object ID before passing the data to parse()? Can we manipulate the queries if they're not efficient?

I don't know that using ORM models will be any simpler than what we're doing now. `PyMongo` can already handle a serialized dict, which pydantic objects provide. And pydantic objects can already parse a dict, which is the output of pymongo queries. What does another layer give us?

Reply:
1. I haven't seen a way to create documents from pydantic models. 
2. IDK
3. You can always write queries if you want

The downside of it is that you have to define the document obviously. The upside is more trivial queries, for example adding a new connection:
```
node.connections.add(connection)
node.save()
```
Since this is not doing any complex queries it's not worth it, but we should keep this option in mind

Reply:
> Since this is not doing any complex queries it's not worth it, but we should keep this option in mind

Agreed. Note, however, that as queries become more complex, they also tend to become less efficient. If efficiency becomes a concern, then you'll need to write your own queries anyway, since the ORM usually won't have the smarts to optimize your query for you.

If you can't tell, I'm very skeptical of the value of ORMs. Here are some opinions on both sides of the issue:

https://blog.codinghorror.com/object-relational-mapping-is-the-vietnam-of-computer-science/
https://martinfowler.com/bliki/OrmHate.html